Python/optimizer_analysis.c

```diff
@@ -1,50 +1,1068 @@
+{
+    if ((sym->types & (1 << typ)) == 0) {
+        return false;
+    }
+    return true;
+}
+
+static inline bool
+sym_matches_type(_Py_UOpsSymType *sym, _Py_UOpsSymExprTypeEnum typ, uint64_t refinement)
+{
+    if (!sym_is_type(sym, typ)) {
+        return false;
+    }
+    if (typ <= MAX_TYPE_WITH_REFINEMENT) {
+        return sym->refinement[typ] == refinement;
+    }
+    return true;
+}
+
+static uint64_t
+sym_type_get_refinement(_Py_UOpsSymType *sym, _Py_UOpsSymExprTypeEnum typ)
+{
+    assert(sym_is_type(sym, typ));
+    assert(typ <= MAX_TYPE_WITH_REFINEMENT);
+    return sym->refinement[typ];
+}
+
+
+static inline bool
+op_is_end(uint32_t opcode)
+{
+    return opcode == _EXIT_TRACE || opcode == _JUMP_TO_TOP;
+}
+
+static inline bool
+op_is_guard(uint32_t opcode)
+{
+    return _PyUop_Flags[opcode] & HAS_GUARD_FLAG;
+}
+
+static inline bool
+op_is_pure(uint32_t opcode)
+{
+    return _PyUop_Flags[opcode] & HAS_PURE_FLAG;
+}
+
+static inline bool
+op_is_bookkeeping(uint32_t opcode) {
+    return (opcode == _SET_IP ||
+            opcode == _CHECK_VALIDITY ||
+            opcode == _SAVE_RETURN_OFFSET ||
+            opcode == _RESUME_CHECK);
+}
+
+static inline bool
+op_is_specially_handled(uint32_t opcode)
+{
+    return _PyUop_Flags[opcode] & HAS_SPECIAL_OPT_FLAG;
+}
+
+static inline bool
+is_const(_Py_UOpsSymType *expr)
+{
+    return expr->const_val != NULL;
+}
+
+static inline PyObject *
+get_const(_Py_UOpsSymType *expr)
+{
+    return expr->const_val;
+}
+
+
+static int
+clear_locals_type_info(_Py_UOpsAbstractInterpContext *ctx) {
+    int locals_entries = ctx->frame->locals_len;
+    for (int i = 0; i < locals_entries; i++) {
+        _Py_UOpsSymType *new_local = sym_init_unknown(ctx);
+        if (new_local == NULL) {
+            return -1;
+        }
+        sym_copy_immutable_type_info(ctx->frame->locals[i], new_local);
+        ctx->frame->locals[i] = new_local;
+    }
+    return 0;
+}
+
+static inline int
+emit_i(uops_emitter *emitter,
+       _PyUOpInstruction inst)
+{
+    if (emitter->curr_i < 0) {
+        OPT_STAT_INC(optimizer_failure_reason_no_writebuffer);
+        DPRINTF(1, "out of emission space\n");
+        return -1;
+    }
+    if (emitter->writebuffer + emitter->curr_i >= emitter->writebuffer_end) {
+        OPT_STAT_INC(optimizer_failure_reason_no_writebuffer);
+        DPRINTF(1, "out of emission space\n");
+        return -1;
+    }
+    if (inst.opcode == _NOP) {
+        return 0;
+    }
+    DPRINTF(2, "Emitting instruction at [%d] op: %s, oparg: %d, target: %d, operand: %" PRIu64 " \n",
+            emitter->curr_i,
+            (inst.opcode >= 300 ? _PyOpcode_uop_name : _PyOpcode_OpName)[inst.opcode],
+            inst.oparg,
+            inst.target,
+            inst.operand);
+    emitter->writebuffer[emitter->curr_i] = inst;
+    emitter->curr_i++;
+    return 0;
+}
+
+static inline int
+emit_const(uops_emitter *emitter,
+       PyObject *const_val,
+       _PyUOpInstruction shrink_stack)
```

Comment:
I wonder if the code wouldn't be more elegant if this function took the number of items to be popped off the stack instead of an instruction. It could then have a local struct which it would initialize using the pop count, rather than there being a single `shrink_stack` struct declared at the top of `uop_abstract_interpret_single_inst()` shared by all the cases. So the generated code would have
```c
                 if (emit_const(&ctx->emitter, (PyObject *)res, NUM_POPS) < 0) { goto error; }
```
instead of
```c
                shrink_stack.oparg = NUM_POPS;
                 if (emit_const(&ctx->emitter, (PyObject *)res, shrink_stack) < 0) { goto error; }
```